From 7dfad42a54afe61470d13fd4e911e9923eb97a89 Mon Sep 17 00:00:00 2001
From: Jeevaka Badrappan <jeevaka.badrappan@intel.com>
Date: Mon, 29 Apr 2013 14:14:37 +0300
Subject: [PORT FROM MAIN] Telephony: Device not registering on flight mode exit

BZ: 85235

When entering Flight mode, in some platforms radio
may be switched off but not the SIM. So, SIM
notifications can be received in flight mode.
Currently, restoring of saved network selection is
done on SIM ready. Since radio is off in flight mode,
restoring network selection request will be rejected.
So, when the flight mode is deactivated, no network
selection request is sent thus resulting in device
not getting registered.

Restore the saved network selection only if the radio is
on and sim is ready.

Change-Id: Ia707f95010bd521647ebac24d971a364f6c08efc
Orig-Change-Id: Icdfcb25aa400810f9fe940927de76a2c3671d282
Signed-off-by: Jeevaka Badrappan <jeevaka.badrappan@intel.com>
Signed-off-by: Arun Ravindran <arun.ravindran@intel.com>
---
 .../telephony/gsm/GsmServiceStateTracker.java      |   28 ++++++++++++++-----
 1 files changed, 20 insertions(+), 8 deletions(-)

diff --git a/src/java/com/android/internal/telephony/gsm/GsmServiceStateTracker.java b/src/java/com/android/internal/telephony/gsm/GsmServiceStateTracker.java
index 04355d6..c32765c 100644
--- a/src/java/com/android/internal/telephony/gsm/GsmServiceStateTracker.java
+++ b/src/java/com/android/internal/telephony/gsm/GsmServiceStateTracker.java
@@ -195,6 +195,7 @@ final class GsmServiceStateTracker extends ServiceStateTracker {
                 (PowerManager)phone.getContext().getSystemService(Context.POWER_SERVICE);
         mWakeLock = powerManager.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, WAKELOCK_TAG);
 
+        mCi.registerForOn(this, EVENT_RADIO_ON, null);
         mCi.registerForAvailable(this, EVENT_RADIO_AVAILABLE, null);
         mCi.registerForRadioStateChanged(this, EVENT_RADIO_STATE_CHANGED, null);
 
@@ -231,6 +232,7 @@ final class GsmServiceStateTracker extends ServiceStateTracker {
     public void dispose() {
         checkCorrectThread();
         // Unregister for all events.
+        mCi.unregisterForOn(this);
         mCi.unregisterForAvailable(this);
         mCi.unregisterForRadioStateChanged(this);
         mCi.unregisterForVoiceNetworkStateChanged(this);
@@ -272,18 +274,28 @@ final class GsmServiceStateTracker extends ServiceStateTracker {
                 //setPowerStateToDesired();
                 break;
 
+            case EVENT_RADIO_ON:
+                if (mUiccApplcation == null
+                        || mUiccApplcation.getState() != AppState.APPSTATE_READY) {
+                    break;
+                }
+                // Fall through to restore saved network selection if sim is ready
             case EVENT_SIM_READY:
 
-                boolean skipRestoringSelection = mPhone.getContext().getResources().getBoolean(
-                        com.android.internal.R.bool.skip_restoring_network_selection);
+                // SIM can be accessible in radio off
+                if (mCi.getRadioState().isOn()) {
+
+                    boolean skipRestoringSelection = mPhone.getContext().getResources().getBoolean(
+                            com.android.internal.R.bool.skip_restoring_network_selection);
 
-                if (!skipRestoringSelection) {
-                    // restore the previous network selection.
-                    mPhone.restoreSavedNetworkSelection(null);
+                    if (!skipRestoringSelection) {
+                        // restore the previous network selection.
+                        mPhone.restoreSavedNetworkSelection(null);
+                    }
+                    pollState();
+                    // Signal strength polling stops when radio is off
+                    queueNextSignalStrengthPoll();
                 }
-                pollState();
-                // Signal strength polling stops when radio is off
-                queueNextSignalStrengthPoll();
                 break;
 
             case EVENT_RADIO_STATE_CHANGED:
-- 
1.7.4.1

